R/edgeThresh.R

```diff
@@ -0,0 +1,284 @@
+    bidirPairList <- vector("list", length = (numBidirectional / 2))
+    for (i in seq(1, numBidirectional, 2)) {
+        bidirPairList[[(i %/% 2) + 1]] <- c(bidirectionalEdges[i], bidirectionalEdges[i + 1])
+    }
+
+    ### For each of the N permutated graphs:
+
+    # Create a new copy of the original EGG graph
+    currentGraph <- EGG
+    ## Swap bidirectional edges first:
+
+    # Randomly pick one from the two nodes involved in a random bidirectional edge,
+    # and swap with another such node, ensure to update both edges involved in a
+    # bidirectional edge
+    i = 0
+    # Swap Q * |E_b| times
+    while (i < (Q * numBidirectional)) {
+        pair <- sample(bidirPairList, size = 2, replace = FALSE)
+        # And then change edge participants
+        # Must delete mutated edges, and add to graph as new edges (even though
+        # the sending nodes are the same)
+        # Use add_edges() and delete_edges()
+        # Can use igraph::any_multiple() to check if multiple edges are created
+        # after each swap, do not increment i if this is the case.
+
+        # Check the new components of all nodes against their previous ones,
+        # using the hash structure created above, discard change if components
+        # are changed (Does this address the disconnected components issue?)
+    }
+
+    # Bidirectional edge swap must not result in multiple edges, thus the new
+    # endpoint nodes are checked for the existence of an already existing edge
+    # from the newly picked node pairs. Neighbor count will remain the same
+    # after a bidirectional edge swap, therefore no need to check.
+
+    # EGG <- random.graph.game(n = 10, p.or.m = 8, type = "gnm", directed = TRUE)
+
+
+    ## Then swap unidirectional edges:
+
+    # At every unidirectional edge swap, receiving endpoints will have a different
+    # number of neighbors, while sending endpoint's node degrees remain the same
+    # Therefore the receiving nodes must have another neighbor that targets them
+    # also do an edge swap, connecting them to another node.
+    # Edge endpoints can be retrieved using:
+    # igraph::ends(EGG, es = E(EGG)[vertices])[,2]
+    # Therefore edge swaps must be done in two successive pairs:
+    # The first pair is selected randomly while the second pair is always the
+    # receiving endpoints of the first pair (!!!Doesn't this result in a chain)?
+
+    # igraph::each_edge() can be used to rewire endpoints, can check to see if
+    # randomly selected edge is a bidirectional edge (which would be already
+    # processed)
+
+    # igraph::keeping_degseq() allows rewiring while preserving "total" degree
+    # distribution (of the graph), not sure if it can be used here
+    # a <- rewire(graph = EGG, with = keeping_degseq())
+
+    # Function degree() with modes: "in", "out" and "all" can be used to check
+    # how many edges point to and from a node.
+
+
+    ### After both kinds of edge swap (or preparation), each node will be
```

Comment:
I don't think a pair swap can split a graph into weakly connected components. Not sure about strongly connected components though.